reconcile/openshift_prometheusrules.py

```diff
@@ -0,0 +1,273 @@
+
+        for slx in pp[category]:
+            if slx['rule'] not in slo_rules:
+                errors.append('Unknown slo rule %s' % slx['rule'])
+
+    # check availability names are unique and slo rules exist
+    availability_rule_names = set()
+    for slx in pp['availability']:
+        availability_rule_names.add(slx['name'])
+        for c in ['latency', 'errors']:
+            for rule_name in slx['rules'][category]:
+                if rule_name not in slos[category]:
+                    errors.append('Unknown %s rule %s' % (category,
+                                                          rule_name))
+
+    if len(availability_rule_names) != len(pp[category]):
+        errors.append('slo names are not unique for %s' % category)
+
+    return errors
+
+
+# Build params to pass to the template
+def build_template_params(pp):
+    params = {}
+    params['http_rates'] = []     # sloRules of http_rate type
+    params['latency_rates'] = []  # sloRules of latency_rate type
+    params['all_rules'] = []      # contains the name of each rule.
+
+    # We have checked that rule names are unique by category, but not as whole
+    # we will add a suffix to each rule to make sure they're unique
+    for r in pp['sloRules']:
+        if r['kind'] == 'http_rate':
+            params['http_rates'].append(r)
+            params['all_rules'].extend(
+                ['%s__http_rates.rateRules' % r['name'],
+                 '%s__http_rates.errorRateRules' % r['name']])
+        else:
+            params['latency_rates'].append(r)
+            params['all_rules'].append('%s__latency_rates.rules' % r['name'])
+
+    for c in ['volume', 'latency', 'errors', 'availability']:
+        params['all_rules'].extend(['%s__%s.rules' % (r['name'], c)
+                                    for r in pp[c]])
+        params[c] = pp[c]
+
+    params['labels'] = pp['labels']
+    params['component'] = pp['component']
+    params['namespace'] = pp['namespace']['name']
+    params['prometheus_labels'] = pp['prometheusLabels']
+
+    return params
+
+
+def fetch_desired_state(performance_parameters, ri):
+    for pp in performance_parameters['performance_parameters_v1']:
+        if pp['namespace']['cluster']['observabilityNamespace'] is None:
+            logging.error('No observability namespace for %s' % pp['name'])
+            continue
+
+        try:
+            errors = check_data_consistency(pp)
+            if len(errors) > 0:
+                logging.error(
+                    'Data inconsistent for %s. Errors detected: %s' % (
+                        pp['name'], errors))
+                continue
+
+            params = build_template_params(pp)
+            rules_resource = generate_resource(SLO_RULES, params)
+        except Exception as e:
+            logging.error('Error building resource for %s: %s' % (
+                pp['name'], e))
+            logging.debug(traceback.format_exc())
+            continue
```

Comment:
shouldn't we fail the integration at this stage? probably not stop execution but make sure we return an errors status. Basically what Maor said about registering errors.

Reply:
Also, I wouldn't make the `try/except` so wide. Can we just limit the try/except to `generate_resource` and catch `utils.jsonnet.JsonnetError`?

What other exceptions are you expecting?

Reply:
I'm trying to avoid that an error in one file affects the rest. I do agree about registering errors.

As for the length of the try/except block, you're right. The errors that can appear in the other parts should be solved in development phase

Reply:
fixed in 21b84a3